nginx/tests/compose.bats

```diff
@@ -0,0 +1,54 @@
+setup() {
+	load 'test_helper/bats-support/load' # this is required by bats-assert
+	load 'test_helper/bats-assert/load'
+
+	set -eu -o pipefail
+}
+
+[[ -n "${DEBUG_BATS:-}" ]] && set -x
+
+@test "API should respond before timeout" {
+	run timeout 15 bash -c "until curl -fksm5 https://localhost:1443; do sleep 0.1; done"
+	assert_success
+}
+
+@test "HTTP request should redirect to HTTPS" {
+	run curl -fsm5 -I http://localhost:1080 
+	assert_success
+	assert_line --partial --index 0 'HTTP/1.1 301 Moved Permanently'
+	assert_line --partial 'Location: https://localhost/'
+}
+
+@test "HTTP acme-challenge should not redirect" {
+	run curl -sm5 -I http://localhost:1080/.well-known/acme-challenge/
+	assert_success
+ 	refute_line --index 0 'HTTP/1.1 301 Moved Permanently'
+	assert_line --partial --index 0 'HTTP/1.1 404 Not Found'
+}
+
+@test "should recieve response from CHT api" {
+	run curl -fksm5 https://localhost:1443/
+	assert_success
+	assert_output 'Hello from CHT api mock'
+
+	run curl -fksm5 https://localhost:1443/somepath
+	assert_success
+	assert_output 'Test'
+
+	run curl -fksm5 -I https://localhost:1443/doesnotexist
+	assert_failure
+	assert_line --partial --index 0 'HTTP/1.1 404 Not Found'
```

Comment:
all tests work for me except this one, specifically the `404` (or maybe something earlier and just fails on last line of `@test`?).  Something I did or a broken test?

```
bats tests/compose.bats                                                                                                                                                                            
 ✓ API should respond before timeout                                                                                                                                                               
 ✓ HTTP request should redirect to HTTPS                                                                                                                                                           
 ✓ HTTP acme-challenge should not redirect                                                                                                                                                         
 ✗ should recieve response from CHT api                                                                                                                                                            
   (from function `assert_line' in file tests/test_helper/bats-assert/src/assert.bash, line 473,                                                                                                   
    in test file tests/compose.bats, line 40)                                                                                                                                                      
     `assert_line --partial --index 0 'HTTP/1.1 404 Not Found'' failed with status 22                                                                                                              
   /home/mrjones/Documents/MedicMobile/cht-core/nginx/tests/test_helper/bats-assert/src/assert.bash: line 473: lines[$idx]: unbound variable                                                       
 ✓ Should return 502 on connection drop                                                                                                                                                            
 ✓ Should return json with on connection drop with accept header                                                                                                                                   
                                                                                                                                                                                                   
6 tests, 1 failure     
```

Reply:
Hmm, this is unexpected, if the test failed to find the string it should give a more helpful error message. Exit code 22 is the expected one from curl with `-f` when encountering 404. Does this test pass?
```bats
@test "should exit" {
	run bash -c 'echo hi ; exit 111'
	assert_failure
	assert_line 'hi'
	assert_line --index 0 'hi'
	assert_line --partial 'hi'
	assert_line --partial --index 0 'hi'
}
```
What version of bash are you using?

Reply:
> Does this test pass? I added that at the top of `./tests/with_mock.bats` and it passes:

```
✓ should exit
```

> What version of bash are you using?

```
bash --version |head -n1
GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)
```

I was able to reproduce the exact same error on a second machine. First was desktop, second was laptop, both Ubuntu 20.04, both bash 5.0.17.

If it's working for you and working in CI on GH, maybe we skip it?

Reply:
Here's my relevant versions:

```
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 
Bats 1.2.0-dev
```

the output of `curl -Ifksm5 https://localhost:1443/doesnotexist` is `NULL`.  Verbose session below. I followed it up with a `-kv` so you could see all the pretty headers and TLS handshakes :eyes: :

```
➜  nginx git:(dn/8179-nginx-json) ✗ make start  
docker-compose -f docker-compose.test.yml up --wait --build
[+] Building 3.4s (14/14) FINISHED                                                                                                                                                                                 
 => [internal] load build definition from Dockerfile                                                                                                                                                          0.0s
 => => transferring dockerfile: 963B                                                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                                                             0.0s
 => => transferring context: 2B                                                                                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/nginx:1.19.6-alpine                                                                                                                                        1.3s
 => [internal] load build context                                                                                                                                                                             0.0s
 => => transferring context: 20.30kB                                                                                                                                                                          0.0s
 => [base_nginx 1/9] FROM docker.io/library/nginx:1.19.6-alpine@sha256:c2ce58e024275728b00a554ac25628af25c54782865b3487b11c21cafb7fabda                                                                       0.1s
 => => resolve docker.io/library/nginx:1.19.6-alpine@sha256:c2ce58e024275728b00a554ac25628af25c54782865b3487b11c21cafb7fabda                                                                                  0.0s
 => => sha256:c2ce58e024275728b00a554ac25628af25c54782865b3487b11c21cafb7fabda 1.65kB / 1.65kB                                                                                                                0.0s
 => => sha256:01747306a7247dbe928db991eab42e4002118bf636dd85b4ffea05dd907e5b66 1.36kB / 1.36kB                                                                                                                0.0s
 => => sha256:629df02b47c8733258baf6663e308a86cd23f80247d35407022c35fd91a50ea3 7.99kB / 7.99kB                                                                                                                0.0s
 => [base_nginx 2/9] RUN apk add --update --no-cache     curl     socat     sed     bash     openssl                                                                                                          1.4s
 => [base_nginx 3/9] COPY ssl-install.sh /docker-entrypoint.d                                                                                                                                                 0.0s
 => [base_nginx 4/9] RUN chmod 755 /docker-entrypoint.d/ssl-install.sh                                                                                                                                        0.5s
 => [base_nginx 5/9] COPY templates /etc/nginx/templates                                                                                                                                                      0.0s 
 => [base_nginx 6/9] COPY nginx.conf /etc/nginx/nginx.conf                                                                                                                                                    0.0s 
 => [base_nginx 7/9] COPY mime.types /etc/nginx/mime.types                                                                                                                                                    0.0s 
 => [base_nginx 8/9] COPY conf.d/ /etc/nginx/conf.d/                                                                                                                                                          0.0s 
 => [base_nginx 9/9] COPY nginx_error_pages/ /usr/share/nginx/nginx_error_pages/                                                                                                                              0.0s
 => exporting to image                                                                                                                                                                                        0.1s
 => => exporting layers                                                                                                                                                                                       0.1s
 => => writing image sha256:0c1bb5699244ba709f5038e75824003e10bdf88ac3f1ccb9f64306908638fdeb                                                                                                                  0.0s
 => => naming to docker.io/library/nginx-cht-nginx                                                                                                                                                            0.0s
[+] Running 2/2
 ⠿ Container cht-api-mock  Healthy                                                                                                                                                                            1.3s
 ⠿ Container cht-nginx     Healthy                                                                                                                                                                            1.3s

➜  nginx git:(dn/8179-nginx-json) ✗ curl -Ifksm5 https://localhost:1443/doesnotexist

➜  nginx git:(dn/8179-nginx-json) ✗ curl -kv https://localhost:1443/doesnotexist
*   Trying 127.0.0.1:1443...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 1443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-SHA
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: emailAddress=domains@medic.org; C=US; ST=Colorado; L=Denver; O=medic; OU=Department of Information Security; CN=test.medic.org
*  start date: Aug 14 20:31:43 2023 GMT
*  expire date: Aug 13 20:31:43 2024 GMT
*  issuer: emailAddress=domains@medic.org; C=US; ST=Colorado; L=Denver; O=medic; OU=Department of Information Security; CN=test.medic.org
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> GET /doesnotexist HTTP/1.1
> Host: localhost:1443
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Server: nginx
< Date: Mon, 14 Aug 2023 20:32:03 GMT
< Content-Length: 0
< Connection: keep-alive
< access-control-allow-origin: 
< access-control-allow-credentials: false
< access-control-allow-methods: 
< access-control-allow-headers: 
< access-control-expose-headers: 
< access-control-max-age: 0
< 
* Connection #0 to host localhost left intact
```

Your note of testing in a container piqued my interest!  I spun up a 20.04 LXD container with [3 security flags enabled](https://ubuntu.com/tutorials/how-to-run-docker-inside-lxd-containers#2-create-lxd-container). After cloning the repo and checking out this branch, `apt install curl bats make jq`, doing the [docker install one liner](https://github.com/docker/docker-install), I get the same error as on my desktop and laptop.  Here's the full bash history in case it helps. First is on bare metal:

```
➜  nginx git:(dn/8179-nginx-json) ✗ lxc launch ubuntu:20.04 nginx-cht-core-8446
➜  nginx git:(dn/8179-nginx-json) ✗ lxc config set nginx-cht-core-8446 security.nesting=true security.syscalls.intercept.mknod=true  security.syscalls.intercept.setxattr=true
➜  nginx git:(dn/8179-nginx-json) ✗ lxc restart nginx-cht-core-8446
➜  nginx git:(dn/8179-nginx-json) ✗ lxc exec nginx-cht-core-8446 -- bash
```

Then is inside the container:

```
root@nginx-cht-core-8446:~/cht-core/nginx# history
git clone https://github.com/medic/cht-core.git
apt update
apt install curl bats make jq
curl -fsSL https://get.docker.com -o get-docker.sh
sh get-docker.sh
docker run hello-world
curl --version
bats -v
cd cht-core/nginx/
git checkout dn/8179-nginx-json
make test
```

I end up with `curl 7.68.0` and `Bats 1.2.0-dev`.   Also, the output of `curl -Ifksm5 https://localhost:1443/doesnotexist` is `NULL` in the container, like was on my laptop and desktop.

Reply:
This is interesting, could you try one or both of the following:
1. removing `-s` from nginx and `--index 0` from the first assert_line
2. removing `-f` from nginx and `assert_failure`

If that doesn't work I'd love to schedule a debug session

Reply:
I'm sorry - I got a little confused on what you'd like me to test.  I think you're talking about the `@test "should exit"` you had me add and you meant `curl` here instead of `nginx`? 

> removing -s from nginx

I think it'd be helpful if you could just give me an updated `with_mock.bats` you want me to run - or do an downstream branch off this PR if you wanna get fancy - either works!

Reply:
>  you meant `curl` here instead of `nginx`?

Yes, my bad, see [8446/files#r1298098028](https://github.com/medic/cht-core/pull/8446/files#r1298098028) for details